Concurrency Issue with Document Lock Implementation

<!-- Please only use this template for submitting common issues -->

**Description**:

After a change to the Document's lock implementation, when many `Document.update()` calls are made at the same time, sometimes `serverSeq` is updated with an outdated value.

This issue can be reproduced using the "0-100-10" test case in `⁠BenchmarkPresenceConcurrency` where ⁠`initialCnt` is set to 0. Since this is a concurrency-related issue, the test passes(or fails) intermittently, so it needs to run a test multiple times to observe the failure.

Interestingly, when `⁠initialCnt` is greater than 0 (when 1+ clients have been attached before calling many `document.Update()` at the same time), the issue doesn't occur. This is why the test cases "100-100-10" and "300-100-10" consistently pass.


Below is the error output from a failed "0-100-10" test:

```
$ go test -tags bench -bench=. ./test/bench/presence_concurrency_bench_test.go -count=10

--- FAIL: BenchmarkPresenceConcurrency/0-100-10-10
    presence_concurrency_bench_test.go:58: 
                Error Trace:    /Users/xo/yorkie/test/bench/presence_concurrency_bench_test.go:58
                                                        /opt/homebrew/opt/go/libexec/src/runtime/asm_arm64.s:1223
                Error:          Received unexpected error:
                                failed_precondition: serverSeq(2) of request greater than serverSeq(0) of ClientInfo: invalid server seq
                Test:           BenchmarkPresenceConcurrency/0-100-10
    presence_concurrency_bench_test.go:58: 
                Error Trace:    /Users/xo/yorkie/test/bench/presence_concurrency_bench_test.go:58
                                                        /opt/homebrew/opt/go/libexec/src/runtime/asm_arm64.s:1223
                Error:          Received unexpected error:
                                failed_precondition: serverSeq(15) of request greater than serverSeq(1) of ClientInfo: invalid server seq
                Test:           BenchmarkPresenceConcurrency/0-100-10
    presence_concurrency_bench_test.go:58: 
                Error Trace:    /Users/xo/yorkie/test/bench/presence_concurrency_bench_test.go:58
                                                        /opt/homebrew/opt/go/libexec/src/runtime/asm_arm64.s:1223
                Error:          Received unexpected error:
                                failed_precondition: serverSeq(9) of request greater than serverSeq(1) of ClientInfo: invalid server seq
                Test:           BenchmarkPresenceConcurrency/0-100-10
    ......
```


**NOTE**

The first concurrency issue appeared in a36cd0e. at this point `⁠conflict on update` error occurred during `⁠Document.update()`.
From ff96799 onward, the error changed to the serverSeq mismatch error shown above, rather than the previous `⁠conflict on update` error.

**Why**:

--- a/server/backend/database/mongo/client.go
+++ b/server/backend/database/mongo/client.go
@@ -854,8 +854,6 @@ func (c *Client) FindOrCreateDocInfo(
 	clientRefKey types.ClientRefKey,
 	docKey key.Key,
 ) (*database.DocInfo, error) {
-	now := gotime.Now()
-
 	filter := bson.M{
 		"project_id": clientRefKey.ProjectID,
 		"key":        docKey,
@@ -863,49 +861,40 @@ func (c *Client) FindOrCreateDocInfo(
 			"$exists": false,
 		},
 	}
-	res, err := c.collection(ColDocuments).UpdateOne(ctx, filter, bson.M{
-		"$set": bson.M{
-			"accessed_at": now,
-		},
-	}, options.Update().SetUpsert(true))
-	if err != nil && !mongo.IsDuplicateKeyError(err) {
-		return nil, fmt.Errorf("upsert document: %w", err)
-	}
 
-	var result *mongo.SingleResult
-	if res != nil && res.UpsertedCount > 0 {
-		result = c.collection(ColDocuments).FindOneAndUpdate(ctx, bson.M{
-			"project_id": clientRefKey.ProjectID,
-			"_id":        res.UpsertedID,
-		}, bson.M{
+	now := gotime.Now()
+	result := c.collection(ColDocuments).FindOneAndUpdate(
+		ctx,
+		filter,
+		bson.M{
 			"$set": bson.M{
+				"accessed_at": now,
+			},
+			"$setOnInsert": bson.M{
 				"owner":      clientRefKey.ClientID,
 				"server_seq": 0,
 				"created_at": now,
 				"updated_at": now,
 			},
-		}, options.FindOneAndUpdate().SetReturnDocument(options.After))
-	} else {
+		},
+		options.FindOneAndUpdate().
+			SetUpsert(true).
+			SetReturnDocument(options.After),
+	)
+
+	if result.Err() != nil && mongo.IsDuplicateKeyError(result.Err()) {
+		// NOTE(hackerwins): If duplicate key error occurred, retry with a
+		// simple find operation since another concurrent request successfully
+		// created the document.
 		result = c.collection(ColDocuments).FindOne(ctx, filter)
-		if result.Err() == mongo.ErrNoDocuments {
-			return nil, fmt.Errorf(
-				"%s %s: %w",
-				clientRefKey.ProjectID,
-				docKey,
-				database.ErrDocumentNotFound,
-			)
-		}
-		if result.Err() != nil {
-			return nil, fmt.Errorf("find document: %w", result.Err())
-		}
 	}
 
-	docInfo := database.DocInfo{}
-	if err := result.Decode(&docInfo); err != nil {
+	info := &database.DocInfo{}
+	if err := result.Decode(info); err != nil {
 		return nil, fmt.Errorf("decode document: %w", err)
 	}
 
-	return &docInfo, nil
+	return info, nil
 }
 
 // FindDocInfoByKey finds the document of the given key.

--- a/test/bench/presence_concurrency_bench_test.go
+++ b/test/bench/presence_concurrency_bench_test.go
@@ -34,40 +34,52 @@ import (
 )
 
 func benchmarkPresenceConcurrency(b *testing.B, svr *server.Yorkie, initialCnt int, concurrentCnt int, syncCnt int) {
-	ctx := context.Background()
-	docKey := key.Key(fmt.Sprintf("presence-concurrency-bench-%d-%d-%d-%d", initialCnt, concurrentCnt, syncCnt, b.N))
+	// Reset the timer to exclude setup time
+	b.ResetTimer()
 
-	// 01. Prepare n clients and attach them to the document.
-	clients, _, err := helper.ClientsAndAttachedDocs(ctx, svr.RPCAddr(), docKey, initialCnt)
-	assert.NoError(b, err)
-	helper.CleanupClients(b, clients)
+	for i := 0; i < b.N; i++ {
+		// Stop the timer during setup
+		b.StopTimer()
 
-	// 02. Then create new clients and attach them to the document concurrently.
-	var wg sync.WaitGroup
-	for range concurrentCnt {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
+		ctx := context.Background()
+		docKey := key.Key(fmt.Sprintf("presence-concurrency-bench-%d-%d-%d-%d", initialCnt, concurrentCnt, syncCnt, i))
 
-			client, doc, err := helper.ClientAndAttachedDoc(ctx, svr.RPCAddr(), docKey)
-			if err != nil {
-				b.Logf("client and doc creation: %v", err)
-				return
-			}
+		// 01. Prepare n clients and attach them to the document.
+		clients, _, err := helper.ClientsAndAttachedDocs(ctx, svr.RPCAddr(), docKey, initialCnt)
+		assert.NoError(b, err)
 
-			assert.NoError(b, client.Sync(ctx))
+		// Start the timer for the actual benchmark operation
+		b.StartTimer()
 
-			for i := range syncCnt {
-				assert.NoError(b, doc.Update(func(root *json.Object, p *presence.Presence) error {
-					p.Set("key", fmt.Sprintf("%d", i))
-					return nil
-				}))
-			}
-			assert.NoError(b, client.Sync(ctx))
-			assert.NoError(b, client.Close())
-		}()
+		// 02. Then create new clients and attach them to the document concurrently.
+		var wg sync.WaitGroup
+		for range concurrentCnt {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				client, doc, err := helper.ClientAndAttachedDoc(ctx, svr.RPCAddr(), docKey)
+				assert.NoError(b, err)
+				err = client.Sync(ctx)
+				assert.NoError(b, err)
+
+				for j := range syncCnt {
+					err := doc.Update(func(root *json.Object, p *presence.Presence) error {
+						p.Set("key", fmt.Sprintf("%d", j))
+						return nil
+					})
+					assert.NoError(b, err)
+				}
+				assert.NoError(b, client.Sync(ctx))
+				assert.NoError(b, client.Close())
+			}()
+		}
+		wg.Wait()
+
+		// Stop the timer during cleanup
+		b.StopTimer()
+		helper.CleanupClients(b, clients)
 	}
-	wg.Wait()
 }
 
 func BenchmarkPresenceConcurrency(b *testing.B) {
@@ -85,8 +97,6 @@ func BenchmarkPresenceConcurrency(b *testing.B) {
 			b.Fatal(err)
 		}
 	})
-
-	b.ResetTimer()
 
 	b.Run("0-100-10", func(b *testing.B) {
 		benchmarkPresenceConcurrency(b, svr, 0, 100, 10)

